schema/schema/generate_schema.go

```diff
@@ -0,0 +1,267 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schema
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"reflect"
+	"strings"
+
+	"knative.dev/pkg/schema/docs"
+)
+
+var (
+	trueVal = true
+
+	zero      = 0.0
+	int8Max   = float64(math.MaxInt8)
+	int16Max  = float64(math.MaxInt16)
+	uint8Max  = float64(math.MaxUint8)
+	uint16Max = float64(math.MaxUint16)
+	uint32Max = float64(math.MaxUint32)
+)
+
+func GenerateForType(t reflect.Type) JSONSchemaProps {
+	s := generateStructSchema(t, true)
+	return s
+}
+
+func generateSchema(t reflect.Type, history ...reflect.Type) JSONSchemaProps {
+	if visited(t, history) {
+		return JSONSchemaProps{
+			Type: "object",
```

Comment:
We could have constants for all those JSON types and formats that are currently repeated across the file.